tests/unit_tests/tools/test_structured.py

```diff
@@ -0,0 +1,208 @@
+
+    assert isinstance(structured_tool_input, StructuredTool)
+    assert (
+        structured_tool_input.args_schema.schema()["properties"]
+        == _MockSchema.schema()["properties"]
+    )
+
+
+def test_tool_lambda_multi_args_schema() -> None:
+    """Test args schema inference when the tool argument is a lambda function."""
+    tool = StructuredTool.from_function(
+        func=lambda tool_input, other_arg: f"{tool_input}{other_arg}",  # type: ignore
+        name="tool",
+        description="A tool",
+    )
+    assert set(tool.args_schema.schema()["properties"]) == {"tool_input", "other_arg"}
+    expected_args = {
+        "tool_input": {"title": "Tool Input"},
+        "other_arg": {"title": "Other Arg"},
+    }
+    assert tool.args_schema.schema()["properties"] == expected_args
+
+
+def test_tool_partial_function_args_schema() -> None:
+    """Test args schema inference when the tool argument is a partial function."""
+
+    def func(tool_input: str, other_arg: str) -> str:
+        return tool_input + other_arg
+
+    with pytest.raises(pydantic.error_wrappers.ValidationError):
+        # We don't yet support args_schema inference for partial functions
+        # so want to make sure we proactively raise an error
+        StructuredTool(
+            name="tool",
+            description="A tool",
+            func=partial(func, other_arg="foo"),
+        )
+
+
+def test_tool_with_kwargs() -> None:
+    """Test functionality when only return direct is provided."""
+
+    @structured_tool(return_direct=True)
+    def search_api(
+        arg_1: float,
+        ping: str = "hi",
+    ) -> str:
+        """Search the API for the query."""
+        return f"arg_1={arg_1}, ping={ping}"
+
+    assert isinstance(search_api, StructuredTool)
+    result = search_api.run(
+        tool_input={
+            "arg_1": 3.2,
+            "ping": "pong",
+        }
+    )
+    assert result == "arg_1=3.2, ping=pong"
+
+    result = search_api.run(
+        tool_input={
+            "arg_1": 3.2,
+        }
+    )
+    assert result == "arg_1=3.2, ping=hi"
+
+
+def test_empty_args_decorator() -> None:
+    """Test inferred schema of decorated fn with no args."""
+
+    @structured_tool
+    def empty_tool_input() -> str:
+        """Return a constant."""
+        return "the empty result"
+
+    assert isinstance(empty_tool_input, StructuredTool)
+    assert empty_tool_input.name == "empty_tool_input"
+    assert empty_tool_input.args_schema.schema()["properties"] == {}
+    assert empty_tool_input.run({}) == "the empty result"
+
+
+def test_nested_pydantic_args() -> None:
+    """Test inferred schema when args are nested pydantic models."""
+    # This is a pattern that is common with FastAPI methods.
+    # If we only parse a dict input but pass the dict
+    # to the function, we are limited only to primitive types
+    # in general.
+
+    class SomeNestedInput(BaseModel):
+        arg2: str
+
+    class SomeInput(BaseModel):
```

Comment:
E.g., for this example, if we had the `.args` example, the json schema would include a `#ref` by default. It seems fine to directly call the args_schema instead